crates/zksync/core/src/vm/runner.rs

```diff
@@ -80,21 +80,27 @@ where
     );
 
     let caller = env.tx.caller;
-    let transact_to = match env.tx.transact_to {
-        TransactTo::Call(to) => to.to_h160(),
+    let (transact_to, nonce) = match env.tx.transact_to {
+        TransactTo::Call(to) => {
+            (to.to_h160(), ZKVMData::new(db, &mut journaled_state).get_tx_nonce(caller))
+        }
         TransactTo::Create(CreateScheme::Create) |
-        TransactTo::Create(CreateScheme::Create2 { .. }) => CONTRACT_DEPLOYER_ADDRESS,
+        TransactTo::Create(CreateScheme::Create2 { .. }) => (
+            CONTRACT_DEPLOYER_ADDRESS,
+            ZKVMData::new(db, &mut journaled_state).get_deploy_nonce(caller),
```

Comment:
I think this deviates from hardhat - I tried searching the zksync hardhart plugin to understand what nonce they use for deployments, and it seems they always use the tx nonces.
The deployment nonce is handled internally by the system contracts.
After all a deployment is still a transaction.

Also I think we don't use this nonce _everywhere_ for deployments, so it would be inconsistent (pretty sure I saw something in the files above that uses tx nonce for a create)